adal/src/telemetry/java/com/microsoft/aad/adal/HttpEvent.java

```diff
@@ -88,19 +91,116 @@ void setRequestIdHeader(final String requestIdHeader) {
     }
 
     /**
+     * Parses and sets the relevant HttpEvent fields given x-ms-clitelem metadata.
+     *
+     * @param xMsCliTelem the value of the x-ms-clitelem header
+     */
+    void setXMsCliTelemData(final String xMsCliTelem) {
+        // if the header isn't present, do nothing
+        if (StringExtensions.isNullOrBlank(xMsCliTelem)) {
+            return;
+        } else {
+            Logger.d(TAG, "Parsing x-ms-clitelem header: " + xMsCliTelem);
+        }
+
+        // split the header based on the delimiter
+        String[] headerSegments = xMsCliTelem.split(",");
+
+        // make sure the header isn't empty
+        if (0 == headerSegments.length) {
+            Logger.w(TAG, "SPE Ring header missing version field.", null, ADALError.X_MS_CLITELEM_VERSION_UNRECOGNIZED);
+            return;
+        }
+
+        // get the version of this header
+        final String headerVersion = headerSegments[0];
+
+        // declare values tracked by this header
+        String errorCode = null;
+        String subErrorCode = null;
+        String tokenAge = null;
+        String speRing = null;
+
+        if (headerVersion.equals("1")) {
+            // The expected delimiter count of the v1 header
+            final int delimCount = 4;
+
+            // Verify the expected format "<version>, <error_code>, <sub_error_code>, <token_age>, <ring>"
+            if (delimCount != xMsCliTelem.length() - xMsCliTelem.replace(",", "").length()) {
```

Comment:
if you check headerSegments.length here, you dont need to do this math. Or did i miss something here?

Reply:
The `String.split(String)` at line 105 can actually return variable length arrays for valid inputs. I check the count of commas in the header value after determining the version of the `x-ms-clitelem` header to permit this. See examples below:

```java
java> "1,0,0,,".split(",").length
java.lang.Integer res0 = 3 // length is 3

java> "1,0,0,1.2,".split(",").length
java.lang.Integer res2 = 4 // length is 4

java> "1,0,0,,I".split(",").length
java.lang.Integer res1 = 5 // length is 5
```

Reply:
That's interesting behavior (different from Objective-C). Is there any way to change this behavior, so it returns as many results as possible with the given delimiter count?

Reply:
Providing a second "limit" value produces more 'predictable' behavior -- this is what I did at line `132` (once the version is known)

Reply:
This check is a bit weird, can we use regex instead? If you want to check before reading thru the segments, you can do a string.match or you can just call string.split with the regex.

Reply:
@weijjia updated to use regex + additional tests in 050d5a66d3bd492efb23c44eade00081f8399e7d